Support for prereleases (#43140)
LLNL/spack: lib/spack/docs/packaging_guide.rst

"snapshot"-like Version Control System (VCS) tag, a VCS branch such as
``v6-16-00-patches``, or a URL specifying a regularly updated snapshot tarball.

.. _version-comparison:

^^^^^^^^^^^^^^^^^^
Version comparison
^^^^^^^^^^^^^^^^^^

Spack imposes a generic total ordering on the set of versions,
independently from the package they are associated with.

Most Spack versions are numeric, a tuple of integers; for example,
``0.1``, ``6.96`` or ``1.2.3.1``. In this very basic case, version
comparison is lexicographical on the numeric components:
``1.2 < 1.2.1 < 1.2.2 < 1.10``.

Spack can also supports string components such as ``1.1.1a`` and
``1.y.0``. String components are considered less than numeric
components, so ``1.y.0 < 1.0``. This is for consistency with
`RPM <https://bugzilla.redhat.com/show_bug.cgi?id=50977>`_. String
components do not have to be separated by dots or any other delimiter.
So, the contrived version ``1y0`` is identical to ``1.y.0``.

Pre-release suffixes also contain string parts, but they are handled

> in a special way. For example ``1.2.3a1`` is parsed as a pre-release

Reviewer: ```suggestion
in a special way. For example ``1.2.3alpha1`` is parsed as a pre-release
```

of the version ``1.2.3``. This allows Spack to order it before the

> actual release: ``1.2.3a1 < 1.2.3``. Spack supports alpha, beta and

Reviewer: ```suggestion
actual release: ``1.2.3alpha1 < 1.2.3``. Spack supports alpha, beta and
```

> release candidate suffixes: ``1.2a1 < 1.2b1 < 1.2rc1 < 1.2``. Any

Reviewer: ```suggestion
release candidate suffixes: ``1.2alpha1 < 1.2beta1 < 1.2rc1 < 1.2``. Any
```

suffix not recognized as a pre-release is treated as an ordinary
string component, so ``1.2 < 1.2-mysuffix``.

Finally, there are a few special string components that are considered
"infinity versions". They include ``develop``, ``main``, ``master``,
``head``, ``trunk``, and ``stable``. For example: ``1.2 < develop``.
These are useful for specifying the most recent development version of
a package (often a moving target like a git branch), without assigning

> a specific version number.

Reviewer: ```suggestion
a specific version number. Infinity versions are not automatically used when determining the latest version of a package unless explicitly required by another package or user.
```

More formally, the order on versions is defined as follows. A version
string is split into a list of components based on delimiters such as
``.`` and ``-`` and string boundaries. The components are split into
the **release** and a possible **pre-release** (if the last component

> is numeric and the second to last is a string ``a``, ``b`` or ``rc``).

Reviewer: ```suggestion
is numeric and the second to last is a string ``alpha``, ``beta`` or ``rc``).
```
